README.txt
= trails

* http://code.google.com/p/twilio-on-rails/

== DESCRIPTION:

Makes developing twilio applications (even) easier in rails.
Support for SMS, twiml MimeType alias, functional test helpers.

== FEATURES/PROBLEMS:


== SYNOPSIS:

 class ApplicationController < ActionController::Base
 ..
	include Trails::Twilio::CallHandling
 ..
 end

== REQUIREMENTS:

* twilio-rails REST library (provided by twilio)

== INSTALL:

* sudo gem install trails

== DEVELOPERS:


== LICENSE:

   Copyright 2010 Hemant Bhanoo

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.


History.txt
=== 1.1.5 / 2010-06-02

* test support for call guid/call status

=== 1.1.4 / 2010-05-09

Backwards-compatible support for environment-specific twilio configuration

=== 1.1.1 / 2010-03-16

Typo fix.

=== 1.1.0 / 2010-03-14

Twilio's SMS API has changed since the beta release (no url specified for handling sms responses).
* Update the send_sms call to reflect that (backwards-incompatible change)
* Start cleaning up the documentation a bit.

=== 1.0.5 / 2010-03-06

* Ugh. accidentally released with unchecked-in code. undoing.

=== 1.0.4 / 2010-03-06

* Clean up logging a bit

=== 1.0.3 / 2010-03-06

* Fix Manifest so all required files get added (!)

=== 1.0.2 / 2010-02-06

* Add support for default twiml layout
* Add example and tests

=== 1.0.1 / 2010-02-06

* Fix basic errors; still no tests :(

=== 1.0.0 / 2010-02-05

* 1 major enhancement

  * Birthday!



twilio-test.rb
require 'ostruct'

module Twilio
  module TestHelper
    def open_session_as_twilio( as_twilio_opts = {}, *args )
      session = open_session( *args )
      modify_session_with_twilio_opts( session, as_twilio_opts )
      session
    end

    def as_twilio( as_twilio_opts = {}, &block )
      if( @integration_session )
        modify_session_with_twilio_opts( @integration_session, as_twilio_opts )
        # end integration test
      elsif( @controller ) # ok we're in a functional test
        # mess with the controller, allowing us to add parameters
        header_modifier = lambda{ |h,o| modify_headers_with_twilio_opts( h, o ) }
        param_modifier = lambda{ |p,o| modify_params_with_twilio_opts( p, o ) }
        @controller.metaclass.send( :define_method, :process_with_twilio_as_caller ) do |request, response|
          # unfortunately we have to reach a little deep into the request here...
          parameters_to_add = {}
          header_modifier.call( request.env, as_twilio_opts )
          param_modifier.call( parameters_to_add, as_twilio_opts )

          # add_parameters
          unless( parameters_to_add.blank? )
            request.instance_variable_set( :@_memoized_query_string, nil ) # cause the query string to be un-memoized
            add_parameters( parameters_to_add )
          end

          process_without_twilio_as_caller( request, response )
        end # def process_with_twilio_as_caller
        @controller.metaclass.send( :alias_method_chain, :process, :twilio_as_caller )

        # need to to easily add parameters
        @controller.metaclass.send( :define_method, :add_parameters ) do |params|
          params ||= {}
          request.query_parameters.merge!( params )
          new_uri = request.request_uri + '&' + params.
            collect{|k,v| "#{CGI::escape(k.to_s)}=#{CGI::escape(v.to_s)}"}.
            join('&')
          request.set_REQUEST_URI( new_uri )
          request.__send__( :instance_variable_set, :@parameters, nil )
        end # add_parameters
      end # functional test

      # cool. call the controller action now:
      block.call()

    end # as_twilio

    def user_presses( digits )
      { 'Digits' => digits }
    end

    def user_records( sound_url )
      { 'RecordingUrl' => sound_url }
    end

    protected
    def modify_session_with_twilio_opts( session, as_twilio_opts )

      header_modifier = lambda{ |h,o| modify_headers_with_twilio_opts( h, o ) }
      param_modifier = lambda{ |p,o| modify_params_with_twilio_opts( p, o ) }

      session.metaclass.send( :define_method, :process_with_twilio_as_caller ) do |method, path, params, headers|
        params ||= {}
        headers ||= {}

        header_modifier.call( headers, as_twilio_opts )
        param_modifier.call( params, as_twilio_opts )

        process_without_twilio_as_caller( method, path, params, headers )
      end # define process_with_twilio_as_caller
      session.metaclass.send( :alias_method_chain, :process, :twilio_as_caller )
    end

    def modify_headers_with_twilio_opts( headers, as_twilio_opts )
      account = if( as_twilio_opts[:account].blank? )
                  cfg = Twilio::Account.send( :config )
                  cfg[cfg.keys.first]
                else
                  as_twilio_opts[:account]
                end
      headers['HTTP_X_TWILIO_ACCOUNTSID'] = account[:sid]
    end

    def modify_params_with_twilio_opts( params, as_twilio_opts )
      caller = as_twilio_opts[:caller] || '4155551212'
      called = as_twilio_opts[:called] || '6155556161'
      params['Caller'] = caller
      params['Called'] = called
    end

    private
    module IntegrationDSL
    end

  end # module TestHelper
end # module Twilio

class TwilioRest::Account
  def self.faked_requests
    return @@fake_requests
  end
  def request_with_fake( url, method, params )
    @@fake_requests ||= []
    @@fake_requests.push( OpenStruct.new( :url => url, :method => method, :params => params ) )
    fake_response = OpenStruct.new
    fake_response.body = 'Fake Body'
    fake_response
  end
  alias_method_chain :request, :fake
end





test/test_trails.rb
require "test/unit"
require "trails"

class TestTrails < Test::Unit::TestCase
  def test_truth
    assert( true )
  end

  

end


lib/trails/test_helper.rb
require 'ostruct'

module Trails
  module TestHelper
    def open_session_as_twilio( as_twilio_opts = {}, *args )
      session = open_session( *args )
      modify_session_with_twilio_opts( session, as_twilio_opts )
      session
    end

    def as_twilio_sms( twilio_opts = {}, &block )
      twilio_opts[:sms] = true
      as_twilio( twilio_opts, &block )
    end

    def as_twilio( as_twilio_opts = {}, &block )
      if( @integration_session )
        modify_session_with_twilio_opts( @integration_session, as_twilio_opts )
        # end integration test
      elsif( @controller ) # ok we're in a functional test
        # mess with the controller, allowing us to add parameters
        header_modifier = lambda{ |h,o| modify_headers_with_twilio_opts( h, o ) }
        param_modifier = lambda{ |p,o| modify_params_with_twilio_opts( p, o ) }
        @controller.metaclass.send( :define_method, :process_with_twilio_as_caller ) do |request, response|
          # unfortunately we have to reach a little deep into the request here...
          parameters_to_add = {}
          header_modifier.call( request.env, as_twilio_opts )
          param_modifier.call( parameters_to_add, as_twilio_opts )

          # add_parameters
          unless( parameters_to_add.blank? )
            request.instance_variable_set( :@_memoized_query_string, nil ) # cause the query string to be un-memoized
            add_parameters( parameters_to_add )
          end

          process_without_twilio_as_caller( request, response )
        end # def process_with_twilio_as_caller
        @controller.metaclass.send( :alias_method_chain, :process, :twilio_as_caller )

        # need to to easily add parameters
        @controller.metaclass.send( :define_method, :add_parameters ) do |params|
          params ||= {}
          request.query_parameters.merge!( params )
          new_uri = request.request_uri + '&' + params.
            collect{|k,v| "#{CGI::escape(k.to_s)}=#{CGI::escape(v.to_s)}"}.
            join('&')
          request.set_REQUEST_URI( new_uri )
          request.__send__( :instance_variable_set, :@parameters, nil )
        end # add_parameters
      end # functional test

      # cool. call the controller action now:
      block.call()

    end # as_twilio
    
    # message defaults to @response.body
    def assert_length_of_sms( message = nil )
      message ||= @response.body
      assert_block( build_message( "SMS should have been no longer than ? characters, but was ? characters.", 
                                   Trails::Twilio::Account::MAX_SMS_LENGTH, message.size ) ) {
        message.size <= Trails::Twilio::Account::MAX_SMS_LENGTH 
      }
    end


    def user_presses( digits )
      { 'Digits' => digits }
    end

    def user_records( sound_url )
      { 'RecordingUrl' => sound_url }
    end

    protected
    def modify_session_with_twilio_opts( session, as_twilio_opts )

      header_modifier = lambda{ |h,o| modify_headers_with_twilio_opts( h, o ) }
      param_modifier = lambda{ |p,o| modify_params_with_twilio_opts( p, o ) }

      session.metaclass.send( :define_method, :process_with_twilio_as_caller ) do |method, path, params, headers|
        params ||= {}
        headers ||= {}

        header_modifier.call( headers, as_twilio_opts )
        param_modifier.call( params, as_twilio_opts )

        process_without_twilio_as_caller( method, path, params, headers )
      end # define process_with_twilio_as_caller
      session.metaclass.send( :alias_method_chain, :process, :twilio_as_caller )
    end

    def modify_headers_with_twilio_opts( headers, as_twilio_opts )
      account = if( as_twilio_opts[:account].blank? )
                  cfg = Twilio::Account.send( :config )
                  cfg[cfg.keys.first]
                else
                  as_twilio_opts[:account]
                end
      headers['HTTP_X_TWILIO_ACCOUNTSID'] = account[:sid]
    end

    def modify_params_with_twilio_opts( params, as_twilio_opts )
      options = as_twilio_opts.dup
      
      options[ 'Caller' ] = options.delete(:caller) || '4155551212'
      options[ 'Called' ] = options.delete(:called) || '6155556161'
      options[ 'From' ]   = options.delete(:from) || '6665554321'
      options[ 'To' ]     = options.delete(:to) || '3334445678'
      options['SmsMessageSid'] = 'DummyMessageSid' if( options[:sms] )

      valid_options = Hash[ *( options.select{ |k,v| 
                                 Trails::Twilio::Incoming.const_get( 'INCOMING_VARS' ).
                                   include?( k ) }.flatten ) ]
      params.merge!( valid_options )
    end

    private
    module IntegrationDSL
    end

  end # module TestHelper
end # module Trails

# open up the TwilioRest::Account class so that we can keep track of faked requests
class TwilioRest::Account
  @@fake_requests ||= []
  def self.faked_requests
    return @@fake_requests
  end
  def request_with_fake( url, method, params )
    @@fake_requests.push( OpenStruct.new( :url => url, :method => method, :params => params ) )
    fake_response = OpenStruct.new
    fake_response.body = 'Fake Body'
    fake_response
  end
  alias_method_chain :request, :fake
end


lib/twiliorest.rb
=begin
Copyright (c) 2008 Twilio, Inc.

Permission is hereby granted, free of charge, to any person
obtaining a copy of this software and associated documentation
files (the "Software"), to deal in the Software without
restriction, including without limitation the rights to use,
copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following
conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.
=end

module TwilioRest
    require 'net/http'
    require 'net/https'
    require 'uri'
    require 'cgi'
    
    TWILIO_API_URL = 'https://api.twilio.com'
    
    class TwilioRest::Account
        
        #initialize a twilio account object
        #
        #id: Twilio account SID/ID
        #token: Twilio account token
        #
        #returns a Twilio account object
        def initialize(id, token)
            @id = id
            @token = token
        end
        
        def _urlencode(params)
            params.to_a.collect! \
                { |k, v| "#{k}=#{CGI.escape(v.to_s)}" }.join("&")
        end
        
        def _build_get_uri(uri, params)
            if params && params.length > 0
                if uri.include?('?')
                    if uri[-1, 1] != '&'
                        uri += '&'
                    end
                    uri += _urlencode(params)
                else
                    uri += '?' + _urlencode(params)
                end
            end
            return uri
        end
        
        def _fetch(url, params, method=nil)
            if method && method == 'GET'
                url = _build_get_uri(url, params)
            end
            uri = URI.parse(url)
            
            http = Net::HTTP.new(uri.host, uri.port)
            http.use_ssl = true
            
            if method && method == 'GET'
                req = Net::HTTP::Get.new(uri.request_uri)
            elsif method && method == 'DELETE'
                req = Net::HTTP::Delete.new(uri.request_uri)
            elsif method && method == 'PUT'
                req = Net::HTTP::Put.new(uri.request_uri)
                req.set_form_data(params)
            else
                req = Net::HTTP::Post.new(uri.request_uri)
                req.set_form_data(params)
            end
            req.basic_auth(@id, @token)
            
            return http.request(req)
        end
        
        #sends a request and gets a response from the Twilio REST API
        #
        #path: the URL (relative to the endpoint URL, after the /v1
        #url: the HTTP method to use, defaults to POST
        #vars: for POST or PUT, a dict of data to send
        #
        #returns Twilio response XML or raises an exception on error
        def request(path, method=nil, vars={})
            if !path || path.length < 1
                raise ArgumentError, 'Invalid path parameter'
            end
            if method && !['GET', 'POST', 'DELETE', 'PUT'].include?(method)
                raise NotImplementedError, 'HTTP %s not implemented' % method
            end
            
            if path[0, 1] == '/'
                uri = TWILIO_API_URL + path
            else
                uri = TWILIO_API_URL + '/' + path
            end
            
            return _fetch(uri, vars, method)
        end
    end
end


lib/trails/twilio/account.rb
module Trails
  module Twilio
    class Account
      attr_reader :config
      def initialize( opts = {} )
        _logger = opts[:logger] || ActiveRecord::Base.logger rescue Logger.new( STDERR )
        if( !opts.blank? ) 
          _logger.warn "overriding default opts #{self.class.config.inspect} with #{opts.inspect}"
        else
          opts = self.class.config[self.class.config.keys.first]
        end
        @config = opts.dup
        @sid = @config[:sid] || raise( "no sid specified on #{self}" )
        @token = @config[:token]
        @logger = _logger
      end

      def self.sid_from_request( request )
        ( :development == RAILS_ENV.to_sym ) ? request.params['AccountSid'] : request.env["HTTP_X_TWILIO_ACCOUNTSID"]
      end

      def self.from_request( request )
        sid = sid_from_request( request )
        unless( config.has_key?( sid ) )
          logger.warn{ "unknown twilio account #{sid}. Request params: #{request.inspect}" }
          raise Trails::Exception::UnknownAccount.new( sid )
        end
        account = new( config[sid].dup )
        raise Trails::Exception::InvalidSignature unless account.verify_caller( request )
        account
      end

      def verify_caller( request )
        # TODO: check caller credentials here. :)
        return true
      end

      # Make outgoing calls:
      # Required:
      # - number
      # - handler_url
      # 
      # Options:
      # - :caller
      # - :method
      # - :timeout
      def call( number, handler_url, opts = {} )
        params = {
          'Caller' => opts['Caller'] || opts[:caller],
          'Called' => number,
          'Url' => handler_url,
          'Method' => opts['Method'] || opts[:method] || 'GET',
          'Timeout' => opts['Timeout'] || opts[:timeout] || 15
        }
        request( 'Calls', 'POST', params )
      end

      MAX_SMS_LENGTH = 160
      # Required:
      # - number:        to
      # - body:          text
      #
      # Options:
      # - :from:       number
      # - :method:     GET/POST
      #
      def send_sms( number, body, opts = {} )
        params = {
          'From' => opts[:from] || @config[:default_number],
          'To' => number,
          'Body' => body,
          'Method' => opts[:method] || 'POST'
        }
        request( 'SMS/Messages', 'POST', params ) 
      end

      # Sample Response: 
      #  [{"SmsFallbackUrl"=>nil, "SmsUrl"=>nil, "PhoneNumber"=>"4253954994", "AccountSid"=>"AC6c25b3d8b4f0a2a4e49e4936398a2180", "Capabilities"=>{"SMS"=>"false", "Voice"=>"false"}, "Method"=>"POST", "Sid"=>"PNc939a026d5a332d22c23ca94a161ce29", "DateUpdated"=>"Sat, 20 Mar 2010 17:52:33 -0700", "DateCreated"=>"Sat, 20 Mar 2010 17:52:33 -0700", "Url"=>nil, "FriendlyName"=>"(425) 395-4994", "VoiceFallbackUrl"=>nil, "SmsFallbackMethod"=>"POST", "VoiceCallerIdLookup"=>"false", "SmsMethod"=>"POST", "VoiceFallbackMethod"=>"POST"}]
      def incoming_numbers( reset = false )
        if( @incoming_numbers.nil? || reset )
          response = 
            request( 'IncomingPhoneNumbers', 'GET' )
          
          if( 200 == response.code.to_i )
            @raw_incoming_numbers = Hash.from_xml( response.body ) 
          else
            raise "got response code #{response.code} and body #{response.body}"
          end
          @incoming_numbers = [@raw_incoming_numbers['TwilioResponse']['IncomingPhoneNumbers']['IncomingPhoneNumber']].flatten # returns an array even when it's a single entry
        end
        return @incoming_numbers
      end

      def outgoing_numbers( reset = false )
        if( @outgoing_numbers.nil? || reset )
          response = 
            request( 'OutgoingCallerIds', 'GET' )
          @outgoing_numbers_raw = Hpricot( response.body ) if( 200 == response.code.to_i )
          @outgoing_numbers = @outgoing_numbers_raw.search( '//phonenumber').
            collect{|j| j.inner_html}
        end
        return @outgoing_numbers
      end

      # options:         [:area_code, :friendly_name, :url, :sms_url]
      # sameple return:
      #    {"SmsFallbackUrl"=>nil,
      #       "SmsUrl"=>nil,
      #       "PhoneNumber"=>"4253954994",
      #       "AccountSid"=>"AC6c25b3d8b4f0a2a4e49e4936398a2180",
      #       "Capabilities"=>{"SMS"=>"false",
      #       "Voice"=>"false"},
      #       "Method"=>"POST",
      #       "Sid"=>"PNc939a026d5a332d22c23ca94a161ce29",
      #       "DateUpdated"=>"Sat,
      #       20 Mar 2010 17:52:33 -0700",
      #       "DateCreated"=>"Sat,
      #       20 Mar 2010 17:52:33 -0700",
      #       "Url"=>nil,
      #       "FriendlyName"=>"(425) 395-4994",
      #       "VoiceFallbackUrl"=>nil,
      #       "SmsFallbackMethod"=>"POST",
      #       "VoiceCallerIdLookup"=>"false",
      #       "SmsMethod"=>"POST",
      #       "VoiceFallbackMethod"=>"POST"}
      def provision_number( options = {} )
        params = {}
        [:area_code, :friendly_name, :url, :sms_url].each do |key|
          params[key.to_s.camelize] = options[key] if options.has_key?( key )
        end

        response = request( 'IncomingPhoneNumbers/Local', 'POST', params )
        if( 201 == response.code.to_i )
          raw_number_response = Hash.from_xml( response.body )
        else
          raise "while trying to acquire a new number, got response #{response.code} and body: #{response.body}"
        end
        raw_number_response["TwilioResponse"]["IncomingPhoneNumber"]
      end

      def release_number( sid )
        request( File.join( 'IncomingPhoneNumbers', sid ), 'DELETE' )
      end

      # just specify the resource (e.g. 'Calls' ) and it will
      # append it to the base uri ("/#{api_version}/Accounts/#{sid}/")
      # and then call twilio.
      def request( resource, method = 'GET', params = {})
        url = File.join( base_uri, resource )
        make_request( url, method, params )
      end
    

      protected

      
      # This makes it easy to create and call the TwilioRest library without
      # having to worry about where credentials come from and stuff.
      def make_request( *args )
        @twilio_account ||= TwilioRest::Account.new( @sid, @token )
        logger.debug{ "making twilio request with #{args.inspect}" }
        @twilio_account.request( *args )
      end

      def base_uri( opts = {} )
        api_version = opts[:api_version] || @api_version || '2008-08-01'
        sid = opts[:sid] || @sid
        "/#{api_version}/Accounts/#{sid}/"
      end

      def logger
        self.class.logger
      end
      def self.logger
        return @logger unless @logger.nil?
        @logger = Logger.new( STDERR )
        @logger.level = Logger::WARN
        return @logger
      end
      def self.config
        @@all_cfg ||= YAML::load_file( config_file ).freeze
        # allow per-environment configuration
        @@cfg ||= if ( @@all_cfg.has_key?( RAILS_ENV ) ) 
                    @@all_cfg[ RAILS_ENV ] 
                  elsif( @@all_cfg.has_key?( 'default' ) )
                    @@all_cfg['default']
                  else
                    @@all_cfg
                  end
      end

      def self.config_file
        return File.join( RAILS_ROOT, 'config', 'twilio.yml' )
      end

    end # class Account
  end # module Twilio
end # module Trails


lib/trails/twilio/incoming.rb
module Trails
  module Twilio
    class Incoming
      attr_reader :account
      def initialize( request, opts = {} )
        @request = request
        @account = Trails::Twilio::Account.from_request( request )
      end

      def twilio_data
        request.params.slice( *INCOMING_VARS ).dup
      end

      def is_sms?
        !sms_message_sid.blank?
      end

      def complete?
        'completed' == self.call_status.downcase
      end

      protected
      attr_reader :request

      INCOMING_VARS = [
        # Always available:
        'CallSid',	# A unique identifier for this call, generated by Twilio.
        'AccountSid',	# Your Twilio account id. It is 34 characters long, and always starts with the letters AC.
        'From',	# The phone number of the party that initiated the call. Formatted with a '+' and country code e.g., +16175551212 (E.164 format). If the call is inbound, then it is the caller's caller ID. If the call is outbound, i.e., initiated via a request to the REST API, then this is the phone number you specify as the caller ID.
        'To',	# The phone number of the called party. Formatted with a '+' and country code e.g., +16175551212 (E.164 format). If the call is inbound, then it's your Twilio phone number. If the call is outbound, then it's the phone number you provided to call.
        'CallStatus',	# A descriptive status for the call. The value is one of queued, ringing, in-progress, completed, busy, failed or no-answer. See the CallStatus section below for more details.
        'ApiVersion',	# The version of the Twilio API used to handle this call. For incoming calls, this is determined by the API version set on the called number. For outgoing calls, this is the API version used by the outgoing call's REST API request.
        'Direction',	# Indicates the direction of the call. In most cases this will be inbound, but if you are using <Dial> it will be outbound-dial.
        'ForwardedFrom',	# This parameter is set only when Twilio receives a forwarded call, but its value depends on the caller's carrier including information when forwarding. Not all carriers support passing this information.
        'FromCity',	# The city of the caller.
        'FromState',	# The state or province of the caller.
        'FromZip',	# The postal code of the caller.
        'FromCountry',	# The country of the caller.
        'ToCity',	# The city of the called party.
        'ToState',	# The state or province of the called party.
        'ToZip',	# The postal code of the called party.
        'ToCountry',	# The country of the called party.

        # Gather:
        'Digits',         # The digits received from the caller

        'RecordingUrl',   # The URL of the recorded audio file
        'Duration', 	# The time duration of the recorded audio file
        'Digits',         # What (if any) key was pressed to end the recording

        # SMS:
        'SmsMessageSid',  # Message SID
        'AccountSid',     # Account ID
        'From',           # 
        'To',             #
        'Body',           # 160 chars

      ].freeze
      public
      INCOMING_VARS.uniq.each do |pname|
        mname = pname.gsub( /[A-Z]/ ) { |s| "_#{s.downcase}" }.gsub( /^_/, '' )
        # some extra debugging code here:
        # ActionController::Base.logger.debug{ "defining method: #{mname} for param #{pname}" }
        define_method( mname ) do
          return request.params[ pname ] || request.env["HTTP_X_TWILIO_#{pname.upcase}"]
        end # define_method
      end # each 

    end #
  end # module Twilio
end # module Trails


twilio.rb
module Twilio
  #
  # = Configuration
  # your config/twilio.yml should look like:
  #
  # "AC1b6a21acb6asdfp9a8sdfj4c3389fb02": # your sid
  # :sid: "AC1b6a21acb6asdfp9a8sdfj4c3389fb02" # your sid again
  # :token: "asdf9ag7742ll32l4kj0dfasdfpa87" # your token
  #
  #
  # = Usage
  # <pre>
  # class ApplicationController < ActionController::Base
  #   include Twilio::CallHandling
  #   ...
  # end
  # </pre>
  #
  # A fake Mime type of 'twiml' will be created.
  # In your actions, you can now do a:
  # <pre>
  # respond_to do |format|
  #   format.twiml { ... }
  # end
  # </pre>
  # 
  # and your templates can be named:
  #  foo/bar.twiml.builder # using builder for template generation is cleanest
  #
  
  module CallHandling
    def self.included( klass )
      raise "can\'t include #{self} in #{klass} - not a Controller?" unless 
        klass.respond_to?( :before_filter )
      Mime::Type.register_alias( "text/html", :twiml ) unless Mime.const_defined?( 'TWIML' )
      klass.send( :before_filter, :setup_incoming_call )
      klass.send( :attr_reader, :incoming_call )
      klass.send( :alias_method_chain, :protect_against_forgery?, :twilio )
    end

    protected

    def protect_against_forgery_with_twilio?
      is_twilio_call? ? false : protect_against_forgery_without_twilio?
    end

    def setup_incoming_call
      return unless is_twilio_call?
      request.format = :twiml
      response.content_type = 'text/xml'
      @incoming_call = Twilio::Incoming.new( request )
    end

    # TODO: Move this onto the request object,
    # it makes more sense to say:
    #
    #    request.is_twilio_call?
    #
    def is_twilio_call?
      return !Twilio::Account.sid_from_request( request ).blank?
    end

    protected

  end # module CallHandling

  class UnknownAccount < Exception; end
  class InvalidSignature < Exception; end

  # Cool.
  class Account
    def initialize( opts = {} )
      if( opts.blank? ) 
        STDERR.puts "no opts specified. trying to pull opts from #{self.class.config.inspect}"
        opts = self.class.config[self.class.config.keys.first]
      end
      @opts = opts.dup
      @sid = @opts[:sid] || raise( "no sid specified on #{self}" )
      @token = @opts[:token]
      @logger = @opts[:logger]
    end

    def self.sid_from_request( request )
      ( :development == RAILS_ENV.to_sym ) ? request.params['AccountSid'] : request.env["HTTP_X_TWILIO_ACCOUNTSID"]
    end

    def self.from_request( request )
      sid = sid_from_request( request )
      unless( config.has_key?( sid ) )	
      	      logger.warn{ "unknown account #{sid}. Request params: #{request.inspect}" }
      	      raise UnknownAccount.new( sid )
      end	      
      account = new( config[sid].dup )
      raise InvalidSignature unless account.verify_caller( request )
    end

    def verify_caller( request )
      # todo: check stuff here.
      return true
    end

    # Make outgoing calls.
    def call( number, handler_url, opts = {} )
      params = {
        'Caller' => opts['Caller'],
        'Called' => number,
        'Url' => handler_url,
        'Method' => opts['Method'] || 'GET',
        'Timeout' => opts['Timeout'] || 15
      }
      api_version = opts[:api_version] || '2008-08-01'
      logger.debug( "Calling twlio with params: #{params.inspect}" )
      make_request( File.join( base_uri, 'Calls' ), 'POST', params )
    end

    # still a WIP:
    def send_sms( number, body, url, opts = {} )
      params = {
        'From' => opts['From'],
        'To' => number,
        'Body' => body,
        'Url' => url,
        'Method' => opts['Method'] || 'POST'
      }
      url =  File.join( base_uri, 'SMS/Messages' )
      logger.debug{ "Calling #{url} with #{params.inspect}" }
      make_request(url, 'POST', params ) 
    end

    def incoming_numbers( reset = false )
      if( @incoming_numbers.nil? || reset )
        response = 
          make_request( File.join( base_uri, 'IncomingPhoneNumbers' ), 'GET' )
        
        if( 200 == response.code.to_i )
          @raw_incoming_numbers = Hpricot( response.body ) 
        else
          raise "got response code #{response.code} and body #{response.body}"
        end
        @incoming_numbers = @raw_incoming_numbers.search( '//phonenumber').
          collect{|j| j.inner_html}
      end
      return @incoming_numbers
    end

    def outgoing_numbers( reset = false )
      if( @outgoing_numbers.nil? || reset )
        response = 
          make_request( "/2008-08-01/Accounts/#{@sid}/OutgoingCallerIds",
                                 'GET' )
        @outgoing_numbers_raw = Hpricot( response.body ) if( 200 == response.code.to_i )
        @outgoing_numbers = @outgoing_numbers_raw.search( '//phonenumber').
          collect{|j| j.inner_html}
      end
      return @outgoing_numbers
    end

    protected

    
    # This makes it easy to create and call the TwilioRest library without
    # having to worry about where credentials come from and stuff.
    def make_request( *args )
      @twilio_account ||= TwilioRest::Account.new( @sid, @token )
      @twilio_account.request( *args )
    end

    def base_uri( opts = {} )
      api_version = opts[:api_version] || @api_version || '2008-08-01'
      sid = opts[:sid] || @sid
      "/#{api_version}/Accounts/#{sid}/"
    end

    def logger
      self.logger
    end
    def self.logger
      return @logger unless @logger.nil?
      @logger = Logger.new( STDERR )
      @logger.level = Logger::WARN
      return @logger
    end
    def self.config
      @@cfg ||= YAML::load_file( config_file )
    end

    def self.config_file
      return File.join( RAILS_ROOT, 'config', 'twilio.yml' )
    end

  end # class Account

  class Incoming
    def initialize( request, opts = {} )
      @request = request
      @account = Twilio::Account.from_request( request )
    end

    protected
    attr_reader :request
    def twilio_data
      @twilio_data ||= requests.params.slice( INCOMING_VARS ).dup
    end

    INCOMING_VARS = [
      # Always available:
      'CallGuid',       # A unique identifier for this call, generated by Twilio. It's 34 characters long, and always starts with the letters CA.
      'Caller',         # The phone number of the party that initiated the call. If the call is inbound, then it is the caller's caller-id. If the call is outbound, i.e., initiated by making a request to the REST Call API, then this is the phone number you specify as the caller-id.
      'Called',         # The phone number of the party that was called. If the call is inbound, then it's your application phone number. If the call is outbound, then it's the phone number you provided to call.
      'AccountGuid',    # Your Twilio account number which is the Twilio Account GUID for the call. It is 34 characters long, and always starts with the letters AC.
      'CallStatus',     # The status of the phone call. The value can be "in-progress", "completed", "busy", "failed" or "no-answer". For a call that was answered and is currently going on, the status would be "in-progress". For a call that couldn't be started because the called party was busy, didn't pick up, or the number dialed wasn't valid: "busy", "no-answer", or "failed" would be returned. If the call finished because the call ended or was hung up, the status would be "completed".
      'CallerCity',	# The city of the caller.
      'CallerState',	# The state or province of the caller.
      'CallerZip',	# The postal code of the caller.
      'CallerCountry',	# The country of the caller.
      'CalledCity',	# The city of the called party.
      'CalledState',	# The state or province of the called party.
      'CalledZip',	# The postal code of the called party.
      'CalledCountry',	# The country of the called party.
      # Gather:
      'Digits',         # The digits received from the caller

      'RecordingUrl',   # The URL of the recorded audio file
      'Duration', 	# The time duration of the recorded audio file
      'Digits',         # What (if any) key was pressed to end the recording
      
      ]
      public
    INCOMING_VARS.uniq.each do |parameter|
      mname = parameter.gsub( /[A-Z]/ ) { |s| "_#{s.downcase}" }.gsub( /^_/, '' )
      # ActionController::Base.logger.debug{ "defining method: #{mname} for param #{parameter}" }
      define_method( mname ) do
        return request.params[ parameter ]
      end # define_method
    end # each 

  end #

end # Twilio
